File: day5/py/main.py
def main():
    file_content = open("input.txt", "r", encoding="utf-8").read()
    
    list = file_content.split("\n")
        
    after_num = {}
    result_list_to_check_mid = []
    
    for elem in list:
        elem_len = len(elem)
        
        if elem_len == 5:
            left = elem[:2] # It would be better to find an index of "|" but anyway :)
            right = elem[3:]
            
            if left in after_num.keys():
               list = after_num[left]
               list.append(right)
            else:
                after_num[left] = [right]
        
        elif elem_len > 5:
            elements_list = elem.split(",")
            elements = elements_list
            elements = elem.split(",")
            elements.reverse()
            
            is_intersection = False
                
            for index, item in enumerate(elements):
                
                elem_to_check = elements[index : index + 1][0]
                list_to_check_in = elements[index + 1:]
                
                elem_list = after_num[elem_to_check]                
                intersection_list = intersection(list_to_check_in, elem_list)
                
                if len(intersection_list) > 0:
                    is_intersection = True
                    break
            
            if is_intersection == False:
                result_list_to_check_mid.append(elements_list)

    
    print("result_list_to_check_mid = ", result_list_to_check_mid)        
        
    sum = 0
    if len(result_list_to_check_mid) > 0:
        for l in result_list_to_check_mid:
            elem_list_len = len(l)
            middle_index = int((elem_list_len - 1)/2)
            middle_elem = int(l[middle_index])
            sum += middle_elem
            
    print("sum = ", sum)
    return

def intersection(lst1, lst2):
    return list(set(lst1) & set(lst2))


main()

File: day5/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	// Read file
	file, err := os.Open("../input.txt")
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	// Read file content
	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	// Initialize maps and slices
	afterNum := make(map[string][]string)
	var resultListToCheckMid [][]string

	// Process each line
	for _, elem := range lines {
		elemLen := len(elem)

		if elemLen == 5 {
			left := elem[:2]
			right := elem[3:]

			if list, exists := afterNum[left]; exists {
				afterNum[left] = append(list, right)
			} else {
				afterNum[left] = []string{right}
			}
		} else if elemLen > 5 {
			elementsList := strings.Split(elem, ",")
			elements := make([]string, len(elementsList))
			copy(elements, elementsList)

			// Reverse elements
			for i, j := 0, len(elements)-1; i < j; i, j = i+1, j-1 {
				elements[i], elements[j] = elements[j], elements[i]
			}

			isIntersection := false

			for index := range elements {
				elemToCheck := elements[index]
				listToCheckIn := elements[index+1:]

				if elemList, exists := afterNum[elemToCheck]; exists {
					intersectionList := intersection(listToCheckIn, elemList)
					if len(intersectionList) > 0 {
						isIntersection = true
						break
					}
				}
			}

			if !isIntersection {
				resultListToCheckMid = append(resultListToCheckMid, elementsList)
			}
		}
	}

	fmt.Println("result_list_to_check_mid =", resultListToCheckMid)

	// Calculate sum
	sum := 0
	if len(resultListToCheckMid) > 0 {
		for _, l := range resultListToCheckMid {
			elemListLen := len(l)
			middleIndex := (elemListLen - 1) / 2
			middleElem, err := strconv.Atoi(strings.TrimSpace(l[middleIndex]))
			if err != nil {
				fmt.Println("Error converting string to integer:", err)
				continue
			}
			sum += middleElem
		}
	}

	fmt.Println("sum =", sum)
}

// intersection returns the intersection of two string slices
func intersection(lst1, lst2 []string) []string {
	// Create a map to store elements from lst1
	set := make(map[string]bool)
	for _, item := range lst1 {
		set[item] = true
	}

	// Create result slice
	var result []string
	for _, item := range lst2 {
		if set[item] {
			result = append(result, item)
		}
	}

	return result
}
